src/libexpr/eval.cc

```diff
@@ -237,10 +238,7 @@ EvalState::EvalState(
            /nix/store while using a chroot store. */
         auto accessor = getFSSourceAccessor();
 
-        auto realStoreDir = dirOf(store->toRealPath(StorePath::dummy));
-        if (settings.pureEval || store->storeDir != realStoreDir) {
-            accessor = settings.pureEval ? storeFS : makeUnionSourceAccessor({accessor, storeFS});
-        }
+        accessor = settings.pureEval ? storeFS.cast<SourceAccessor>() : makeUnionSourceAccessor({accessor, storeFS});
```

Comment:
I don't think this is not quite right --- evidently we need better unit tests to catch this.

#14080 removes redundant branches while keeping the `store->storeDir != realStoreDir` fix.

Reply:
No, this change is necessary. Otherwise, if `pureEval` isn't enabled, `rootFS` ends up not containing `storeFS`, so the mounts on `storeFS` won't be visible in `rootFS`.